File: github_cpp/19/121.cpp
#include <iostream>

using namespace std;

int partition(int *a, int low, int high)
{
	int i = low - 1;
	int j = low;
	int pivot = a[high];
	int tmp;
	while(j < high) {
		if (a[j] <= pivot) {
			i ++;
		  tmp = a[i];
		 	a[i] = a[j];
		  a[j] = tmp;	
		}
		j ++;
	}
	i++;
	tmp = a[i];
	a[i] = a[j];
	a[j] = tmp;
	return i;
}

void quickSort(int * a, int low, int high)
{
	if (low < high) {
		int i = partition(a, low, high);
		quickSort(a, low, i - 1);
		quickSort(a, i + 1, high);
	}
}

void printArr(int * a, int len)
{
	for (int i = 0; i < len; ++i) {
		cout << a[i] << "\t";
	}
}

int main(void) {
	int a[] = {2, 10, 18, 7, 6, 49, 38, 65, 97, 76, 13, 27};
	int len = sizeof(a) / sizeof(a[0]);
	cout << "Before Sorting: ";
	printArr(a, len);
	cout << "\nAfter sorting: ";
	quickSort(a, 0, len - 1);
	printArr(a, len);
	return 0;
}

File: github_java_sort/8/121.java
package quickSort;



public class Quicksort {
	
	
public static int partition(int arr[], int low, int high)
 {
     int pivot = arr[high]; 
     int i = (low-1); 
     for (int j=low; j<high; j++)
     {
         
         
         if (arr[j] <= pivot)
         {
             i++;

             
             int temp = arr[i];
             arr[i] = arr[j];
             arr[j] = temp;
         }
     }

     
     int temp = arr[i+1];
     arr[i+1] = arr[high];
     arr[high] = temp;

     return i+1;
 }


 
 public static void sort(int arr[], int low, int high)
 {
     if (low < high)
     {
         
         int pi = partition(arr, low, high);

         
         
         sort(arr, low, pi-1);
         sort(arr, pi+1, high);
     }
 }

 
 public static void printArray(int arr[])
 {
     int n = arr.length;
     for (int i=0; i<n; ++i)
         System.out.print(arr[i]+" ");
     System.out.println();
 }
 
 
 public static double findMedian(int arr[]) {
	 int length = arr.length;
	 if (length % 2 == 1) {
		 
		 return (double)arr[length/2];
	 }
	 
	 
	 return (double)(arr[length/2-1] + arr[(length/2)])/2;
 }

 
 public static void main(String args[])
 {
     int arr[] = {10, 7, 8, 4, 9, 1, 5};
     int n = arr.length;


     sort(arr, 0, n-1);

     System.out.println("sorted array");
     printArray(arr);
     
     System.out.println("median");
     System.out.println(findMedian(arr));
 }
}



	

